### PR TITLE
(PUP-7681) Add test for binary data in catalog

### PR DESCRIPTION
**Pull request by johnduarte**
This commit adds an acceptance test to for a catalog containing
binary data. The integrity of the binary data should remain intact
through the process of transmission from the server and application
on the client.

### CODE REVIEW

**Comment by johnduarte** [comment]
@joshcooper this test should effectively test that the fallback to a pson payload occurs to ensure the integrity of the binary data. I did not explicitly add a check for the `text/pson` content type in the payload. That seems like an implementation detail that is necessary to ensure that the binary data preserved.

**Comment by geoffnichols** [comment]
Per conversation in standup, this should be retargeted at 5.0.x.

File: acceptance/tests/catalog_with_binary_data.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    end
+    # note - master teardown is registered by #mk_tmp_environment_with_teardown
+  end
+
+  step "Create module with binary data file on master" do
+    on(master, "mkdir -p #{environmentpath}/#{tmp_environment}/modules/#{test_num}/{manifests,files}")
+    master_module_manifest = "#{environmentpath}/#{tmp_environment}/modules/#{test_num}/manifests/init.pp"
+    master_module_binary_file = "#{environmentpath}/#{tmp_environment}/modules/#{test_num}/files/binary_data"
+
+    bin_content = '\xC0\xFF'
```
Comment (MosesMendoza) [inline acceptance/tests/catalog_with_binary_data.rb]: @johnduarte I think we want this to be wrapped in double quotes, not single quotes. If wrapped in single quotes, ruby doesn't evaluate the escape sequence to evaluate as hex value.:
```ruby
[8] pry(main)> bin = "\xC0\xFF"
=> "\xC0\xFF"
[9] pry(main)> bin.valid_encoding?
=> false
[10] pry(main)> bin = '\xC0\xFF'
=> "\\xC0\\xFF"
[11] pry(main)> bin.valid_encoding?
=> true
```

Comment (MosesMendoza) [inline acceptance/tests/catalog_with_binary_data.rb]: @johnduarte I think we want this to be wrapped in double quotes, not single quotes. If wrapped in single quotes, ruby doesn't evaluate the escape sequence to evaluate as hex value.:
```ruby
[8] pry(main)> bin = "\xC0\xFF"
=> "\xC0\xFF"
[9] pry(main)> bin.valid_encoding?
=> false
[10] pry(main)> bin = '\xC0\xFF'
=> "\\xC0\\xFF"
[11] pry(main)> bin.valid_encoding?
=> true
```

File: acceptance/tests/catalog_with_binary_data.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  end
+
+  step "start the master" do
+    with_puppet_running_on(master, {}) do
+
+      step "run puppet and ensure that binary data was correctly applied" do
+        agents.each do |agent|
+          on(agent, puppet('agent', '--test', "--environment #{tmp_environment}", "--server #{master.hostname}"), :acceptable_exit_codes => 2)
+          on(agent, "#{Puppet::Acceptance::CommandUtils::ruby_command(agent)} -e 'puts File.binread(\"#{agent_tmp_dirs[agent.hostname]}/#{test_num}\").bytes'") do |res|
+            assert_match(/192\n255/, res.stdout, 'Binary file did not contain originally specified data')
```
Comment (MosesMendoza) [inline acceptance/tests/catalog_with_binary_data.rb]: Though if there was an issue I would expect this assertion to fail... hmm.
```
[10] pry(main)> bin = '\xC0\xFF'
=> "\\xC0\\xFF"
[11] pry(main)> bin.valid_encoding?
=> true
[12] pry(main)> bin.bytes
=> [92, 120, 67, 48, 92, 120, 70, 70]
```

Comment (MosesMendoza) [inline acceptance/tests/catalog_with_binary_data.rb]: I wonder if during the translation to remote call we're effectively unescaping the extra escape backslash added when using single quotes.

File: acceptance/tests/catalog_with_binary_data.rb (hunk 1 above)
Comment (johnduarte) [inline acceptance/tests/catalog_with_binary_data.rb]: hrm.
When I use double quotes, `binwrite` reports that the encoding is invalid.
```
    ArgumentError: invalid byte sequence in UTF-8
```

Using single quotes, the `binwrite`, and the test succeeds as expected.

Tested on RHEL-7 and Windows 2012

**Comment by johnduarte** [comment]
@joshcooper thank you for taking the time to dive into the Beaker implementation that was impacting this. I have incorporated your suggestions and validated that the test passes on centos 7 and windows 2012.

**Pull request merged**